distutils/extension.py

```diff
@@ -106,16 +108,22 @@ def __init__(
     ):
         if not isinstance(name, str):
             raise AssertionError("'name' must be a string")  # noqa: TRY004
-        if not (
-            isinstance(sources, list)
-            and all(isinstance(v, (str, os.PathLike)) for v in sources)
-        ):
+
+        # we handle the string case first; though strings are iterable, we disallow them
+        if isinstance(sources, str):
+            raise AssertionError(  # noqa: TRY004
+                "'sources' must be an iterable of strings or PathLike objects, not a string"
+            )
```

Comment:
Is it possible instead to leverage the Type Annotations to perform the type validation at run time? Feels redundant to be performing these checks manually. e.g. `@typeguard.typechecked` on `__init__` and add an appropriate annotation for it (and name also)?

Reply:
I it possible? technically yes. The amount of support may also depend on the Python version for introspection of annotations on a function, forward references of annotations, stringified annotations, etc.

This is what a library like [pydantic](https://pypi.org/project/pydantic/) is dedicated to do, as an example.

Reply:
Sure, I could add an annotated check, though `typeguard` isn't part of the stdlib – is it? 

To add something like `TypeGuard[Iterable]` with a custom validator that wraps this logic, that also means adding a dependency on `typing_extensions` for Python <=3.10, IIRC,

Reply:
> To add something like `TypeGuard[Iterable]` with a custom validator that wraps this logic, that also means adding a dependency on `typing_extensions` for Python <=3.10, IIRC.

`typing.TypeGuard` and `typing.TypeIs` only make sense as return annotations, and I don't think would be used as part of any runtime validation. So as long as you use `from __future__ import annotations` (or quote the annotation), and lock the `typing_extensions` import behind a `TYPE_CHECKING` block, then you can use `typing_extensions.TypeGuard` on any non-EOL python version.

As for adding a dependency on the library `typeguard`, it would indeed also add a dependency on `typing_extensions` (and `importlib_metadata` for Python 3.9)

Reply:
Thanks for the correction, yes, my idea was to separate the runtime validation from the development-time type hints but you explained it more concretely than I did. :)

I'll wait for further directions, leaving it to @jaraco and you to suggest if opting for `typeguard` as an optional dependency guarded by `TYPE_CHECKING` would be the best way for `distutils` going forward. 

If there's something specific besides this that I can do within this PR, please let me know!

Reply:
I'd like to use typeguard here. I'm okay with adding dependencies, especially when that means avoiding repetitive type-checking logic that's been automated elsewhere. However, it doesn't need to happen for this PR. I'll track that effort separately. Thanks for the background.